include/RAJA/BoxSegment.hxx

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) 2016, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ *
+ * All rights reserved.
+ *
+ * This source code cannot be distributed without permission and
+ * further review from Lawrence Livermore National Laboratory.
+ */
+
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   RAJA header file defining box segment classes.
+ *     
+ ******************************************************************************
+ */
+
+#ifndef RAJA_BoxSegment_HXX
+#define RAJA_BoxSegment_HXX
+
+#include "config.hxx"
+
+#include "BaseSegment.hxx"
+
+#include <algorithm>
+#include <iosfwd>
+
+
+namespace RAJA {
+
+
+/*!
+ ******************************************************************************
+ *
+ * \brief  Segment class representing a Box indices.
+ *
+ *         Box is specified by a corner, number of dimensions,
+ *         extent in each dimentsion, and stride to advance in each dimension.
+ *
+ ******************************************************************************
+ */
+class BoxSegment : public BaseSegment
+{
+public:
+
+   ///
+   /// Default box segment ctor.
+   ///
+   /// Segment undefined until begin/end values set.
+   ///
+   BoxSegment()
+   : BaseSegment( _BoxSeg_ ), 
+     m_corner(UndefinedValue), 
+     m_dim(UndefinedValue) { ; }
+
+   ///
+   /// Construct box segment with [begin, end) specified.
+   ///
+   BoxSegment(Index_type corner,  Index_type dim,
+                Index_type const *extent, Index_type const *stride) 
+   : BaseSegment( _BoxSeg_ ) 
+   {
+      m_corner = corner ;
+      m_dim = dim ;
+
+      for (int i=0; i<m_dim; ++i) {
+         m_extent[i] = extent[i] ;
+         m_stride[i] = stride[i] ;
+      }
+
+      /* eliminate trivially defined dimensions */
+      for (int i = m_dim-1; i>0; --i) {
+         if (m_extent[i] == 1) {
+            if (i != m_dim-1) {
+               for (int j=i; j<m_dim-1; ++j) {
+                  m_extent[i] = m_extent[i+1] ;
+                  m_stride[i] = m_stride[i+1] ;
+               }
+            }
+            --m_dim ;
+         }
+      }
+
+      /* collapse adjacent dimensions with regular stride */
+      for (int i=0; i<(m_dim-1); ++i) {
```

Comment:
This is a very good idea.  It might be worth looking at doing the mixed compile/runtime shape approach we talked about though.  Kokkos actually does this so it can evaluate as much of the shape of its multidimensional array references as it can at compile time before the optimizer gets at it.